iothub_client/src/iothubtransport_mqtt_common.c

```diff
@@ -909,59 +916,154 @@ static int publish_device_method_message(MQTTTRANSPORT_HANDLE_DATA* transport_da
         {
-            MQTT_MESSAGE_HANDLE mqtt_get_msg = mqttmessage_create(mqtt_info->packet_id, STRING_c_str(msg_topic), DELIVER_AT_MOST_ONCE, NULL, 0);
-            if (mqtt_get_msg == NULL)
+            if (mqtt_client_publish(transport_data->mqttClient, mqtt_get_msg) != 0)
             {
-                LogError("Failed constructing mqtt message.");
-                free(mqtt_info);
+                LogError("Failed publishing to mqtt client.");
                 result = __FAILURE__;
             }
             else
             {
-                if (mqtt_client_publish(transport_data->mqttClient, mqtt_get_msg) != 0)
-                {
-                    LogError("Failed publishing to mqtt client.");
-                    free(mqtt_info);
-                    result = __FAILURE__;
-                }
-                else
+                DList_InsertTailList(&transport_data->ack_waiting_queue, &mqtt_info->entry);
+                result = 0;
+            }
+            mqttmessage_destroy(mqtt_get_msg);
+        }
+        STRING_delete(msg_topic);
+    }
+
+    return result;
+}
+
+static void sendPendingGetTwinRequests(PMQTTTRANSPORT_HANDLE_DATA transportData)
+{
+    PDLIST_ENTRY dev_twin_item = transportData->pending_get_twin_queue.Flink;
+
+    while (dev_twin_item != &transportData->pending_get_twin_queue)
+    {
+        DLIST_ENTRY saveListEntry;
+        saveListEntry.Flink = dev_twin_item->Flink;
+
+        MQTT_DEVICE_TWIN_ITEM* msg_entry = containingRecord(dev_twin_item, MQTT_DEVICE_TWIN_ITEM, entry);
+
+        (void)DList_RemoveEntryList(dev_twin_item);
+
+        if (publish_device_twin_get_message(transportData, msg_entry) != 0)
+        {
+            LogError("Failed sending pending get twin request");
+            destroy_device_twin_get_message(msg_entry);
+        }
+
+        dev_twin_item = saveListEntry.Flink;
+    }
+}
+
+static void removeExpiredPendingGetTwinRequests(PMQTTTRANSPORT_HANDLE_DATA transport_data)
+{
+    tickcounter_ms_t current_ms;
+
+    if (tickcounter_get_current_ms(transport_data->msgTickCounter, &current_ms) == 0)
+    {
+        PDLIST_ENTRY listItem = transport_data->pending_get_twin_queue.Flink;
+
+        while (listItem != &transport_data->pending_get_twin_queue)
+        {
+            DLIST_ENTRY nextListItem;
+            nextListItem.Flink = listItem->Flink;
+
+            MQTT_DEVICE_TWIN_ITEM* msg_entry = containingRecord(listItem, MQTT_DEVICE_TWIN_ITEM, entry);
+
+            if (((current_ms - msg_entry->msgEnqueueTime) / 1000) >= ON_DEMAND_GET_TWIN_REQUEST_TIMEOUT_SECS)
+            {
+                (void)DList_RemoveEntryList(listItem);
+                msg_entry->userCallback(DEVICE_TWIN_UPDATE_COMPLETE, NULL, 0, msg_entry->userContext);
+                destroy_device_twin_get_message(msg_entry);
+            }
+
+            listItem = nextListItem.Flink;
+        }
+    }
+}
+
+static void removeExpiredGetTwinRequestsPendingAck(PMQTTTRANSPORT_HANDLE_DATA transport_data)
```

Comment:
>removeExpiredGetTwinRequestsPendingAck [](start = 12, length = 38)

This method isn't being called anywhere. Is this used to drain the ack_waiting_queue on transport destroy?

Reply:
There is a bunch of code to support the expiry of twin requests.  If we don't need this function then let's remove some of that code.

---
In reply to: [242079502](https://github.com/Azure/azure-iot-sdk-c/pull/771#discussion_r242079502) [](ancestors = 242079502)

Reply:
This was supposed to be called within doWork(). Just added. Thanks for pointing out.